sklearn/cross_validation.py

```diff
@@ -700,63 +711,95 @@ def __len__(self):
+                                 "than the number of samples %d" %
+                                 (train_size, n))
+            test_size = n - train_size
+        else:
+            raise ValueError("Invalid value for train_size: %r" % train_size)
+    else:  # test_size is not None and train_set is not None:
+        test_size_kind = np.array(test_size).dtype.kind
+        train_size_kind = np.array(train_size).dtype.kind
+        if test_size_kind == 'f' and \
+           train_size_kind == 'f':
+            if test_size >= 1.:
+                raise ValueError('test_size=%f should be smaller '
+                                 'than 1.0 or be an integer' % test_size)
+            if train_size >= 1.:
+                raise ValueError('train_size=%f should be smaller '
+                                 'than 1.0 or be an integer' % train_size)
+            if train_size + test_size > 1.:
                 raise ValueError('The sum of test_size and train_size = %f, '
                                  'should be smaller than 1.0. Reduce '
                                  'test_size and/or train_size.' %
                                  (train_size + test_size))
-        elif np.asarray(train_size).dtype.kind == 'i':
+        elif test_size_kind == 'i' and train_size_kind == 'i':
+            if test_size >= n:
+                raise ValueError('test_size=%d should be smaller '
+                                 'than the number of samples %d' %
+                                 (test_size, n))
             if train_size >= n:
                 raise ValueError("train_size=%d should be smaller "
                                  "than the number of samples %d" %
                                  (train_size, n))
+            if train_size + test_size > n:
+                raise ValueError('The sum of test_size and train_size = %d, '
+                                 'should be smaller than n. Reduce '
+                                 'test_size and/or train_size.' %
+                                 (train_size + test_size))
         else:
-            raise ValueError("Invalid value for train_size: %r" % train_size)
+            # test_size_kind != train_size_kind or
+            # their kinds are nether 'f' nor 'i'
+            if test_size_kind not in ['f', 'i']:
+                raise ValueError("Invalid value for test_size: %r" %
+                                 test_size)
+            if train_size_kind not in ['f', 'i']:
+                raise ValueError("Invalid value for train_size: %r" %
+                                 train_size)
+            raise ValueError(
+                "Type of test_size and train_size is not the same, "
+                "test_size: %r, train_size: %r" %
+                (type(test_size), type(train_size)))
 
     if np.asarray(test_size).dtype.kind == 'f':
         n_test = ceil(test_size * n)
     elif np.asarray(test_size).dtype.kind == 'i':
         n_test = float(test_size)
```

Comment:
I don't get the need for this line. Does removing the else clauses break any tests?